<a href="https://colab.research.google.com/github/Anishabyte/demo/blob/main/credit_card_fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing the dependencies


In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [9]:
# loading the dataset to a pandas Dataframe
credit_card_data=pd.read_csv('/content/creditcard.csv')

In [10]:
# first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [11]:
credit_card_data.tail()
# printing the last five rows of the dataset

Time        V1        V2        V3        V4        V5        V6  \
107661  70549  1.198279 -0.851404 -0.172063 -0.760588 -0.732620 -0.390914   
107662  70549 -4.608890 -2.417203 -1.684811  0.206299  1.240726  3.368097   
107663  70550 -1.125806  0.163657  1.873904  1.407388  0.400394  0.167562   
107664  70550 -0.423253  0.752399  1.577846  0.280206 -0.094335 -0.263924   
107665  70550  0.907958 -0.923497  0.730426  0.211849 -0.948813  0.409957   

              V7        V8        V9  ...       V21       V22       V23  \
107661 -0.308317 -0.108760 -1.116133  ... -0.317584 -0.724378 -0.145397   
107662 -1.802464  1.883831 -1.202476  ... -0.379691 -0.212864 -0.474509   
107663  0.488385  0.250106 -0.798662  ...  0.117333  0.190644  0.003970   
107664  0.474498  0.088235  0.119206  ...  0.051250  0.557695 -0.004066   
107665 -0.555812  0.199092  1.260090  ... -0.149574 -0.492247 -0.083512   

             V24       V25       V26       V27       V28  Amount  Class  
107661 -0.317131  0.286789  1.054526 -0.093112  0.006122  118.00    0.0  
107662  1.133652 -0.053589 -0.203666  0.022797 -1.153237   12.00    0.0  
107663  0.234093  0.445414 -0.163892  0.050617  0.081551   98.35    0.0  
107664  0.468395 -0.584390  0.284416  0.423211  0.241590    9.20    0.0  
107665 -0.351753  0.064257  0.988019 -0.048000       NaN     NaN    NaN  

[5 rows x 31 columns]

In [12]:
# dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107666 entries, 0 to 107665
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    107666 non-null  int64  
 1   V1      107666 non-null  float64
 2   V2      107666 non-null  float64
 3   V3      107666 non-null  float64
 4   V4      107666 non-null  float64
 5   V5      107666 non-null  float64
 6   V6      107666 non-null  float64
 7   V7      107666 non-null  float64
 8   V8      107666 non-null  float64
 9   V9      107666 non-null  float64
 10  V10     107666 non-null  float64
 11  V11     107666 non-null  float64
 12  V12     107666 non-null  float64
 13  V13     107666 non-null  float64
 14  V14     107666 non-null  float64
 15  V15     107666 non-null  float64
 16  V16     107666 non-null  float64
 17  V17     107666 non-null  float64
 18  V18     107666 non-null  float64
 19  V19     107666 non-null  float64
 20  V20     107666 non-null  float64
 21  V21     10

In [13]:
# checking the number of misssing value in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       1
Amount    1
Class     1
dtype: int64

In [14]:
# dropping all the missing values
credit_card_data=credit_card_data.dropna()

In [15]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [18]:
# checking the distribution of legit transaction and fradulent transaction
credit_card_data['Class'].value_counts()

0.0    107428
1.0       237
Name: Class, dtype: int64

This data is highly unbalaced

0----->Normal transaction
1----->Fradulent transaction

In [19]:
# separating the data for analysis
legit=credit_card_data[credit_card_data.Class==0]
fraud=credit_card_data[credit_card_data.Class==1]

In [20]:
print(legit.shape)
print(fraud.shape)

(107428, 31)
(237, 31)


In [22]:
# stastical measure of the data
legit.Amount.describe()

count    107428.000000
mean         96.131142
std         260.844741
min           0.000000
25%           7.100000
50%          25.170000
75%          86.970000
max       19656.530000
Name: Amount, dtype: float64

In [23]:
fraud.Amount.describe()

count     237.000000
mean      118.885148
std       255.864211
min         0.000000
25%         1.000000
50%         7.610000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [24]:
# compare the values for both transaction
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    44326.693739 -0.244646 -0.035961  0.700676  0.146649 -0.275017   
1.0    38578.008439 -6.020609  4.209147 -7.728592  4.759216 -4.298181   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.101216 -0.102301  0.055409 -0.049402  ...  0.043082 -0.033876   
1.0   -1.547669 -6.396113  1.618176 -2.755420  ...  0.252535  1.368462   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0.0   -0.106727 -0.037017  0.01058  0.133236  0.025663  0.000488  0.001429   
1.0   -0.344519 -0.105012 -0.11118  0.215603  0.092395  0.533794  0.066186   

           Amount  
Class              
0.0     96.131142  
1.0    118.885148  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal transaction and fraudulent transactions

Number of fradulent transaction is 237

In [25]:
legit_sample=legit.sample(n=492)

Concatinating two dataframe

In [26]:
new_dataset=pd.concat([legit_sample,fraud],axis=0)

In [27]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
62549  50313 -0.933480  0.370584  2.380690  0.806094 -1.256009  1.119959   
42168  40978 -0.675169  1.383807  0.542348  0.948633 -0.363193 -0.856115   
74654  55687  1.253583  0.107155 -0.007781  0.332624 -0.365354 -1.251624   
11911  20525  1.145316  0.071595  0.827779  1.230429 -0.371442  0.355653   
46095  42638 -1.160903  1.151422 -0.918765  0.872705 -0.085144 -0.277674   

             V7        V8        V9  ...       V21       V22       V23  \
62549 -0.767293  0.912573  1.284643  ...  0.271730  1.149106 -0.111641   
42168  0.365628  0.233094 -0.403687  ...  0.138380  0.391091  0.018021   
74654  0.287560 -0.296029  0.079325  ... -0.059514 -0.215625 -0.064891   
11911 -0.668756  0.278263  1.628965  ...  0.005147  0.214609 -0.056842   
46095  0.202334  0.899959 -0.761588  ...  0.221992  0.617456  0.039441   

            V24       V25       V26       V27       V28  Amount  Class  
62549  0.151708 -0.116496 -0.300782 -0.060142 -0.018001   53.00    0.0  
42168  0.381528 -0.083928 -0.330524 -0.043626  0.019685   18.90    0.0  
74654  0.429353  0.495357  0.599883 -0.077577  0.004861   27.83    0.0  
11911 -0.413966  0.361166 -0.289520  0.015357  0.003739    3.00    0.0  
46095 -0.309553 -0.723957 -0.370495  0.412846  0.013050   89.99    0.0  

[5 rows x 31 columns]

In [28]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
105178  69394  1.140431  1.134243 -1.429455  2.012226  0.622800 -1.152923   
106679  70071 -0.440095  1.137239 -3.227080  3.242293 -2.033998 -1.618415   
106998  70229  0.315642  1.636778 -1.519650  4.028571 -1.186794 -0.789813   
107067  70270 -1.512516  1.133139 -1.601052  2.813401 -2.664503 -0.310371   
107637  70536 -2.271755 -0.457655 -2.589055  2.230778 -4.278983  0.388610   

              V7        V8        V9  ...       V21       V22       V23  \
105178  0.221159  0.037372  0.034486  ... -0.367136 -0.891627 -0.160578   
106679 -3.028013  0.764555 -1.801937  ...  0.764187 -0.275578 -0.343572   
106998 -2.279807  0.472988 -1.657635  ...  0.345921 -0.108002 -0.165442   
107067 -1.520895  0.852996 -1.496495  ...  0.729828  0.485286  0.567005   
107637  0.102485  0.813128 -1.092921  ...  1.096342  0.658399  1.711676   

             V24       V25       V26       V27       V28  Amount  Class  
105178 -0.108326  0.668374 -0.352393  0.071993  0.113684    1.00    1.0  
106679  0.233085  0.606434 -0.315433  0.768291  0.459623  227.30    1.0  
106998  0.279895  0.808783  0.117363  0.589595  0.309064    3.79    1.0  
107067  0.323586  0.040871  0.825814  0.414482  0.267265  318.11    1.0  
107637  0.333540  0.538591 -0.193529  0.258194  0.247269  824.83    1.0  

[5 rows x 31 columns]

In [29]:
new_dataset['Class'].value_counts()

0.0    492
1.0    237
Name: Class, dtype: int64

In [31]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    41195.428862 -0.250206 -0.036615  0.613645  0.097205 -0.258968   
1.0    38578.008439 -6.020609  4.209147 -7.728592  4.759216 -4.298181   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.170005 -0.212102  0.057319  0.014821  ...  0.062665 -0.003914   
1.0   -1.547669 -6.396113  1.618176 -2.755420  ...  0.252535  1.368462   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.115249 -0.018019  0.016348  0.129741 -0.004582  0.038872  0.016600   
1.0   -0.344519 -0.105012 -0.111180  0.215603  0.092395  0.533794  0.066186   

           Amount  
Class              
0.0     94.839268  
1.0    118.885148  

[2 rows x 30 columns]

Splitting the data into features and target

In [33]:
X=new_dataset.drop('Class',axis=1)
Y=new_dataset['Class']

In [35]:
print(X)

         Time        V1        V2        V3        V4        V5        V6  \
62549   50313 -0.933480  0.370584  2.380690  0.806094 -1.256009  1.119959   
42168   40978 -0.675169  1.383807  0.542348  0.948633 -0.363193 -0.856115   
74654   55687  1.253583  0.107155 -0.007781  0.332624 -0.365354 -1.251624   
11911   20525  1.145316  0.071595  0.827779  1.230429 -0.371442  0.355653   
46095   42638 -1.160903  1.151422 -0.918765  0.872705 -0.085144 -0.277674   
...       ...       ...       ...       ...       ...       ...       ...   
105178  69394  1.140431  1.134243 -1.429455  2.012226  0.622800 -1.152923   
106679  70071 -0.440095  1.137239 -3.227080  3.242293 -2.033998 -1.618415   
106998  70229  0.315642  1.636778 -1.519650  4.028571 -1.186794 -0.789813   
107067  70270 -1.512516  1.133139 -1.601052  2.813401 -2.664503 -0.310371   
107637  70536 -2.271755 -0.457655 -2.589055  2.230778 -4.278983  0.388610   

              V7        V8        V9  ...       V20       V21       V22  \


In [36]:
print(Y)

62549     0.0
42168     0.0
74654     0.0
11911     0.0
46095     0.0
         ... 
105178    1.0
106679    1.0
106998    1.0
107067    1.0
107637    1.0
Name: Class, Length: 729, dtype: float64


Splitting the data into training and testing data

In [38]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [39]:
print(X.shape,X_train.shape,X_test.shape)

(729, 30) (583, 30) (146, 30)


Model Training

Logistic Regression

In [40]:
model=LogisticRegression()

In [41]:
# training the logistic regression model with training data
model.fit(X_train,Y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evalution

Accuracy score

In [42]:
# accuracy on training data
X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)

In [43]:
print('Accuracy on training data:',training_data_accuracy)

Accuracy on training data: 0.9554030874785592


In [45]:
# accuracy on test data
X_test_prediction=model.predict(X_test)
test_data_accuracy=accuracy_score(X_test_prediction,Y_test)

In [46]:
print('accuracy score on test data:',test_data_accuracy)

accuracy score on test data: 0.9041095890410958
